src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,308 @@
+     *    iii> If no MRRT exists, try to find FRT. 
+     * 4) Use FRT:
+     *    i>   If FRT exists, use the FRT. 
+     *    ii>  If FRT request fails, and we haven't tried with MRRT, use MRRT. 
+     *    iii> If FRT does not exist, if we haven't tried with MRRT, use the MRRT. 
+     *    VI>  If FRT exists, and we've already tried with MRRT, return the MRRT result. 
+     */
+    AuthenticationResult getAccessToken() throws AuthenticationException {
+        // If mTokenCacheAccessor is null, won't handle with token cache lookup. 
+        if (mTokenCacheAccessor == null) {
+            return null;
+        }
+        
+        // Check for if there is valid access token item in the cache.
+        final TokenCacheItem accessTokenItem = mTokenCacheAccessor.getATFromCache(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUserFromRequest());
+        if (accessTokenItem == null) {
+            Logger.v(TAG, "No valid access token exists, try with refresh token.");
+            return tryRT();
+        }
+        
+        Logger.v(TAG, "Return AT from cache.");
+        return AuthenticationResult.createResult(accessTokenItem);
+    }
+    
+    /**
+     * Send token request with grant_type as refresh_token to token endpoint for getting new access token. 
+     */
+    AuthenticationResult acquireTokenWithRefreshToken(final String refreshToken) 
+            throws AuthenticationException {
+        Logger.v(TAG, "Try to get new access token with the found refresh token.", 
+                mAuthRequest.getLogInfo(), null);
+        
+        // Check if network is available, if not throw exception. 
+        HttpWebRequest.throwIfNetworkNotAvaliable(mContext);
+        
+        final AuthenticationResult result;
+        try {
+            final JWSBuilder jwsBuilder = new JWSBuilder();
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+            result = oauthRequest.refreshToken(refreshToken);
+            if (result != null && StringExtensions.IsNullOrBlank(result.getRefreshToken())) {
+                Logger.i(TAG, "Refresh token is not returned or empty", "");
+                result.setRefreshToken(refreshToken);
+            }
+        } catch (final IOException | AuthenticationException exc) {
+            // Server side error or similar
+            Logger.e(TAG, "Error in refresh token for request:" + mAuthRequest.getLogInfo(),
+                    ExceptionExtensions.getExceptionMessage(exc), ADALError.AUTH_FAILED_NO_TOKEN,
+                    exc);
+
+            throw new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    exc);
+        }
+
+        return result;
+    }
+    
+    /**
+     * For testing purpose, inject the dependency. 
+     */
+    void setWebRequestHandler(final IWebRequestHandler webRequestHandler) {
```

Comment:
As I mentioned in the TODO(at the beginning of this file), i'm going to remove it, but not in this PR. 

---
In reply to: [63409795](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63409795) [](ancestors = 63409795)